Type cast error in commit observation logic when offsets are pre-batched

### Versions used 

Alpakka Kafka 2.1.0-M1

Akka version: n/a

### Expected Behavior

When the user pre-batches offsets and uses `CommitWhen.NextOffsetObserved` the latest offset for each partition should be deferred.


### Actual Behavior

A type cast runtime exception occurs because Commit When expects single offsets (`CommittableOffset` vs `CommittableOffsetBatch`)


### Relevant logs

```
[ERROR] [18:36:52.431] ak.ac.RepointableActorRef:159 Error in stage [akka.kafka.internal.CommitCollectorStage@25e3cedb]: class akka.kafka.internal.CommittableOffsetBatchImpl cannot be cast to class akka.kafka.ConsumerMessage$CommittableOffset (akka.kafka.internal.CommittableOffsetBatchImpl and akka.kafka.ConsumerMessage$CommittableOffset are in unnamed module of loader 'app')
java.lang.ClassCastException: class akka.kafka.internal.CommittableOffsetBatchImpl cannot be cast to class akka.kafka.ConsumerMessage$CommittableOffset (akka.kafka.internal.CommittableOffsetBatchImpl and akka.kafka.ConsumerMessage$CommittableOffset are in unnamed module of loader 'app')
        at akka.kafka.internal.CommitObservationLogic.updateBatch(CommitObservationLogic.scala:35)
        at akka.kafka.internal.CommitObservationLogic.updateBatch$(CommitObservationLogic.scala:31)
        at akka.kafka.internal.CommitCollectorStageLogic.updateBatch(CommitCollectorStage.scala:35)
        at akka.kafka.internal.CommitCollectorStageLogic.akka$kafka$internal$CommitCollectorStageLogic$$consume(CommitCollectorStage.scala:59)
        at akka.kafka.internal.CommitCollectorStageLogic$$anon$1.onPush(CommitCollectorStage.scala:90)
        at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:541)
        at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:423)
```


### Reproducible Test Case

See PR

--- a/core/src/main/mima-filters/2.1.0-M1.backwards.excludes/PR1262-committables-used-by-commit-when.excludes
+++ b/core/src/main/mima-filters/2.1.0-M1.backwards.excludes/PR1262-committables-used-by-commit-when.excludes
@@ -1,0 +1,1 @@
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.kafka.ConsumerMessage#CommittableOffsetBatch.filter")

--- a/core/src/main/scala/akka/kafka/ConsumerMessage.scala
+++ b/core/src/main/scala/akka/kafka/ConsumerMessage.scala
@@ -219,6 +219,9 @@ object ConsumerMessage {
     @InternalApi
     private[kafka] def tellCommitEmergency(): CommittableOffsetBatch
 
+    @InternalApi
+    private[kafka] def filter(p: GroupTopicPartition => Boolean): CommittableOffsetBatch
+
     /**
      * @return true if the batch contains no commits.
      */

--- a/core/src/main/scala/akka/kafka/internal/CommitObservationLogic.scala
+++ b/core/src/main/scala/akka/kafka/internal/CommitObservationLogic.scala
@@ -32,19 +32,42 @@ private[internal] trait CommitObservationLogic { self: GraphStageLogic =>
     if (settings.when == OffsetFirstObserved) {
       offsetBatch = offsetBatch.updated(committable)
     } else { // CommitWhen.NextOffsetObserved
-      val offset = committable.asInstanceOf[CommittableOffset]
-      val gtp = offset.partitionOffset.key
-      deferredOffsets.get(gtp) match {
-        case Some(dOffset: CommittableOffset) if dOffset.partitionOffset.offset < offset.partitionOffset.offset =>
-          deferredOffsets = deferredOffsets + (gtp -> offset)
-          offsetBatch = offsetBatch.updated(dOffset)
-        case None =>
-          deferredOffsets = deferredOffsets + (gtp -> offset)
-        case _ => ()
+      committable match {
+        case single: CommittableOffset =>
+          val gtp = single.partitionOffset.key
+          updateBatchForPartition(gtp, single, single.partitionOffset.offset)
+        case batch: CommittableOffsetBatchImpl =>
+          for { (gtp, offsetAndMetadata) <- batch.offsetsAndMetadata } updateBatchForPartition(
+            gtp,
+            batch,
+            offsetAndMetadata.offset()
+          )
+        case unknownBatchImpl: CommittableOffsetBatch =>
+          throw new IllegalArgumentException(
+            s"Unknown CommittableOffsetBatch, got [${unknownBatchImpl.getClass.getName}], " +
+            s"expected [${classOf[CommittableOffsetBatchImpl].getName}]"
+          )
+
       }
     }
     offsetBatch.batchSize >= settings.maxBatch
   }
+
+  private def updateBatchForPartition(gtp: GroupTopicPartition, committable: Committable, offset: Long): Unit =
+    deferredOffsets.get(gtp) match {
+      case Some(dOffset: CommittableOffset) if dOffset.partitionOffset.offset < offset =>
+        deferredOffsets = deferredOffsets + (gtp -> committable)
+        offsetBatch = offsetBatch.updated(dOffset)
+      case Some(dOffsetBatch: CommittableOffsetBatch)
+          if dOffsetBatch.offsets.contains(gtp) && dOffsetBatch.offsets
+            .get(gtp)
+            .head < offset =>
+        deferredOffsets = deferredOffsets + (gtp -> committable)
+        offsetBatch = offsetBatch.updated(dOffsetBatch)
+      case None =>
+        deferredOffsets = deferredOffsets + (gtp -> committable)
+      case _ => ()
+    }
 
   /**
    * Clear any deferred offsets and return the count before emptied. This should only be called

--- a/core/src/main/scala/akka/kafka/internal/CommitObservationLogic.scala
+++ b/core/src/main/scala/akka/kafka/internal/CommitObservationLogic.scala
@@ -39,7 +39,7 @@ private[internal] trait CommitObservationLogic { self: GraphStageLogic =>
         case batch: CommittableOffsetBatchImpl =>
           for { (gtp, offsetAndMetadata) <- batch.offsetsAndMetadata } updateBatchForPartition(
             gtp,
-            batch,
+            batch.filter(_.equals(gtp)),
             offsetAndMetadata.offset()
           )
         case unknownBatchImpl: CommittableOffsetBatch =>

--- a/core/src/main/scala/akka/kafka/internal/MessageBuilder.scala
+++ b/core/src/main/scala/akka/kafka/internal/MessageBuilder.scala
@@ -20,7 +20,6 @@ import org.apache.kafka.clients.consumer.{ConsumerRecord, OffsetAndMetadata}
 import org.apache.kafka.common.TopicPartition
 import org.apache.kafka.common.requests.OffsetFetchResponse
 
-import scala.collection.compat._
 import scala.compat.java8.FutureConverters.FutureOps
 import scala.concurrent.Future
 import scala.jdk.CollectionConverters._
@@ -243,6 +242,12 @@ private[kafka] final class CommittableOffsetBatchImpl(
       committers.head._2.tellCommit(this, emergency = emergency)
     }
     this
+  }
+
+  override private[kafka] def filter(p: GroupTopicPartition => Boolean): CommittableOffsetBatch = {
+    val newOffsets = offsetsAndMetadata.filter { case (gtp, _) => p(gtp) }
+    val newCommitters = offsets.map { case (gtp, _) => gtp -> committerFor(gtp) }
+    new CommittableOffsetBatchImpl(newOffsets, newCommitters, newOffsets.size)
   }
 
   override def commitJavadsl(): CompletionStage[Done] = commitInternal().toJava

--- a/core/src/main/scala/akka/kafka/internal/MessageBuilder.scala
+++ b/core/src/main/scala/akka/kafka/internal/MessageBuilder.scala
@@ -247,7 +247,7 @@ private[kafka] final class CommittableOffsetBatchImpl(
   override private[kafka] def filter(p: GroupTopicPartition => Boolean): CommittableOffsetBatch = {
     val newOffsets = offsetsAndMetadata.filter { case (gtp, _) => p(gtp) }
     val newCommitters = offsets.map { case (gtp, _) => gtp -> committerFor(gtp) }
-    new CommittableOffsetBatchImpl(newOffsets, newCommitters, newOffsets.size)
+    new CommittableOffsetBatchImpl(newOffsets, newCommitters, newOffsets.size.toLong)
   }
 
   override def commitJavadsl(): CompletionStage[Done] = commitInternal().toJava

--- a/core/src/main/scala/akka/kafka/internal/MessageBuilder.scala
+++ b/core/src/main/scala/akka/kafka/internal/MessageBuilder.scala
@@ -20,6 +20,7 @@ import org.apache.kafka.clients.consumer.{ConsumerRecord, OffsetAndMetadata}
 import org.apache.kafka.common.TopicPartition
 import org.apache.kafka.common.requests.OffsetFetchResponse
 
+import scala.collection.compat._
 import scala.compat.java8.FutureConverters.FutureOps
 import scala.concurrent.Future
 import scala.jdk.CollectionConverters._

--- a/tests/src/test/scala/akka/kafka/internal/CommitCollectorStageSpec.scala
+++ b/tests/src/test/scala/akka/kafka/internal/CommitCollectorStageSpec.scala
@@ -251,7 +251,7 @@ class CommitCollectorStageSpec(_system: ActorSystem)
 
         control.shutdown().futureValue shouldBe Done
       }
-      "only commit when the next offset is observed in a CommittableOffsetBatch" in {
+      "only commit when the next offset is observed in a CommittableOffsetBatch" in assertAllStagesStopped {
         val (sourceProbe, control, sinkProbe, offsetFactory) = streamProbesWithOffsetFactory(settings)
         // create batches of size 1
         val (batch1, batch2, batch3) =
@@ -280,7 +280,7 @@ class CommitCollectorStageSpec(_system: ActorSystem)
 
         control.shutdown().futureValue shouldBe Done
       }
-      "only commit when the next offset is observed in a CommittableOffset preceded by a CommittableOffsetBatch" in {
+      "only commit when the next offset is observed in a CommittableOffset preceded by a CommittableOffsetBatch" in assertAllStagesStopped {
         val (sourceProbe, control, sinkProbe, offsetFactory) = streamProbesWithOffsetFactory(settings)
         // create a mix of single offsets and batches of 1
         val (batch1, msg2, batch3) =
@@ -305,7 +305,7 @@ class CommitCollectorStageSpec(_system: ActorSystem)
 
         control.shutdown().futureValue shouldBe Done
       }
-      "only commit when the next offset is observed in a CommittableOffsetBatch preceded by a CommittableOffset" in {
+      "only commit when the next offset is observed in a CommittableOffsetBatch preceded by a CommittableOffset" in assertAllStagesStopped {
         val (sourceProbe, control, sinkProbe, offsetFactory) = streamProbesWithOffsetFactory(settings)
         // create a mix of single offsets and batches of 1
         val (msg1, batch2, msg3) =

--- a/tests/src/test/scala/akka/kafka/internal/CommitCollectorStageSpec.scala
+++ b/tests/src/test/scala/akka/kafka/internal/CommitCollectorStageSpec.scala
@@ -10,7 +10,7 @@ import java.util.concurrent.atomic.AtomicLong
 import akka.Done
 import akka.actor.ActorSystem
 import akka.event.LoggingAdapter
-import akka.kafka.ConsumerMessage.{CommittableOffset, CommittableOffsetBatch, PartitionOffset}
+import akka.kafka.ConsumerMessage.{Committable, CommittableOffset, CommittableOffsetBatch, PartitionOffset}
 import akka.kafka.scaladsl.{Committer, Consumer}
 import akka.kafka.testkit.ConsumerResultFactory
 import akka.kafka.testkit.scaladsl.{ConsumerControlFactory, Slf4jToAkkaLoggingAdapter}
@@ -251,6 +251,89 @@ class CommitCollectorStageSpec(_system: ActorSystem)
 
         control.shutdown().futureValue shouldBe Done
       }
+      "only commit when the next offset is observed in a CommittableOffsetBatch" in {
+        val (sourceProbe, control, sinkProbe, offsetFactory) = streamProbesWithOffsetFactory(settings)
+        // create batches of size 1
+        val (batch1, batch2, batch3) =
+          (offsetFactory.makeBatchOffset(), offsetFactory.makeBatchOffset(), offsetFactory.makeBatchOffset())
+
+        sinkProbe.request(100)
+
+        // commit in first batch should not be committed but 'batched-up' again
+        sourceProbe.sendNext(batch1)
+        sourceProbe.sendNext(batch2)
+        sourceProbe.sendNext(batch3)
+
+        val batches = sinkProbe.expectNextN(2)
+        sinkProbe.expectNoMessage(10.millis)
+
+        // batches are committed using mapAsyncUnordered, so it's possible to receive batch acknowledgements
+        // downstream out of order
+        val lastBatch = batches.maxBy(_.offsets.values.last)
+
+        lastBatch.offsets.values.last shouldBe batch2
+          .asInstanceOf[CommittableOffsetBatch]
+          .offsets
+          .head
+          ._2 withClue "expect only the second offset to be committed"
+        offsetFactory.committer.commits.size shouldBe 2 withClue "expected only two commits"
+
+        control.shutdown().futureValue shouldBe Done
+      }
+      "only commit when the next offset is observed in a CommittableOffset preceded by a CommittableOffsetBatch" in {
+        val (sourceProbe, control, sinkProbe, offsetFactory) = streamProbesWithOffsetFactory(settings)
+        // create a mix of single offsets and batches of 1
+        val (batch1, msg2, batch3) =
+          (offsetFactory.makeBatchOffset(), offsetFactory.makeOffset(), offsetFactory.makeBatchOffset())
+
+        sinkProbe.request(100)
+
+        // first batch should not be committed but 'batched-up' again
+        sourceProbe.sendNext(batch1)
+        sourceProbe.sendNext(msg2)
+        sourceProbe.sendNext(batch3)
+
+        val batches = sinkProbe.expectNextN(2)
+        sinkProbe.expectNoMessage(10.millis)
+
+        // batches are committed using mapAsyncUnordered, so it's possible to receive batch acknowledgements
+        // downstream out of order
+        val lastBatch = batches.maxBy(_.offsets.values.last)
+
+        lastBatch.offsets.values.last shouldBe msg2.partitionOffset.offset withClue "expect only the second offset to be committed"
+        offsetFactory.committer.commits.size shouldBe 2 withClue "expected only two commits"
+
+        control.shutdown().futureValue shouldBe Done
+      }
+      "only commit when the next offset is observed in a CommittableOffsetBatch preceded by a CommittableOffset" in {
+        val (sourceProbe, control, sinkProbe, offsetFactory) = streamProbesWithOffsetFactory(settings)
+        // create a mix of single offsets and batches of 1
+        val (msg1, batch2, msg3) =
+          (offsetFactory.makeOffset(), offsetFactory.makeBatchOffset(), offsetFactory.makeOffset())
+
+        sinkProbe.request(100)
+
+        // first batch should not be committed but 'batched-up' again
+        sourceProbe.sendNext(msg1)
+        sourceProbe.sendNext(batch2)
+        sourceProbe.sendNext(msg3)
+
+        val batches = sinkProbe.expectNextN(2)
+        sinkProbe.expectNoMessage(10.millis)
+
+        // batches are committed using mapAsyncUnordered, so it's possible to receive batch acknowledgements
+        // downstream out of order
+        val lastBatch = batches.maxBy(_.offsets.values.last)
+
+        lastBatch.offsets.values.last shouldBe batch2
+          .asInstanceOf[CommittableOffsetBatch]
+          .offsets
+          .head
+          ._2 withClue "expect only the second offset to be committed"
+        offsetFactory.committer.commits.size shouldBe 2 withClue "expected only two commits"
+
+        control.shutdown().futureValue shouldBe Done
+      }
       "only commit when the next offset is observed for the correct partitions" in assertAllStagesStopped {
         val (sourceProbe, control, sinkProbe, offsetFactory) = streamProbesWithOffsetFactory(settings)
         val (msg1, msg2, msg3, msg4, msg5) = (offsetFactory.makeOffset(partitionNum = 1),
@@ -299,12 +382,12 @@ class CommitCollectorStageSpec(_system: ActorSystem)
 
   private def streamProbes(
       committerSettings: CommitterSettings
-  ): (TestPublisher.Probe[CommittableOffset], Consumer.Control, TestSubscriber.Probe[CommittableOffsetBatch]) = {
+  ): (TestPublisher.Probe[Committable], Consumer.Control, TestSubscriber.Probe[CommittableOffsetBatch]) = {
 
     val flow = Committer.batchFlow(committerSettings)
 
     val ((source, control), sink) = TestSource
-      .probe[CommittableOffset]
+      .probe[Committable]
       .viaMat(ConsumerControlFactory.controlFlow())(Keep.both)
       .via(flow)
       .toMat(TestSink.probe)(Keep.both)
@@ -315,7 +398,7 @@ class CommitCollectorStageSpec(_system: ActorSystem)
 
   private def streamProbesWithOffsetFactory(
       committerSettings: CommitterSettings
-  ): (TestPublisher.Probe[CommittableOffset],
+  ): (TestPublisher.Probe[Committable],
       Consumer.Control,
       TestSubscriber.Probe[CommittableOffsetBatch],
       TestOffsetFactory) = {
@@ -346,6 +429,10 @@ class CommitCollectorStageSpec(_system: ActorSystem)
 
     def makeOffset(failWith: Option[Throwable] = None, partitionNum: Int = 1): CommittableOffset = {
       TestCommittableOffset(offsetCounter, committer, failWith, partitionNum)
+    }
+
+    def makeBatchOffset(failWith: Option[Throwable] = None, partitionNum: Int = 1): CommittableOffsetBatch = {
+      CommittableOffsetBatch(makeOffset(failWith, partitionNum))
     }
   }
 

